xla/service/cpu/tests/onednn_convolution_test.cc

```diff
@@ -32,159 +33,216 @@ limitations under the License.
 namespace xla {
 namespace cpu {
 
-class ConvolutionTest : public HloTestBase {
+class ConvolutionTest : public HloTestBase,
+                        public ::testing::WithParamInterface<PrimitiveType> {
  protected:
   DebugOptions GetDebugOptionsForTest() override {
     DebugOptions debug_options = HloTestBase::GetDebugOptionsForTest();
     debug_options.set_xla_cpu_use_thunk_runtime(false);
     return debug_options;
   }
 
-  const char* conv_rewrite_str_ = R"(
-    ; CHECK:     custom_call_target="__onednn$convolution",
-    ; CHECK:       backend_config={
-    ; CHECK-DAG:     "outer_dimension_partitions":[],
-    ; CHECK-DAG:       "onednn_conv_config":{
-    ; CHECK-DAG:   }
-    ; CHECK:     }
-    )";
+  PrimitiveType dtype_;
+  std::string dtypeString_;
+  bool user_scratchpad_;
+  bool weights_prepacked_;
+  float atol_;
+  float rtol_;
 
-  const char* conv_rewrite_bias_str_ = R"(
-    ; CHECK:     custom_call_target="__onednn$convolution",
+  constexpr static char* kConvRewriteStr = R"(
+    ; CHECK:     custom_call_target="__onednn$$convolution",
     ; CHECK:       backend_config={
     ; CHECK-DAG:     "outer_dimension_partitions":[],
-    ; CHECK-DAG:       "onednn_conv_config":{
-    ; CHECK-DAG:       "fusions":{
-    ; CHECK-DAG:         "ops":["BIAS"]
-    ; CHECK-DAG:     }
-    ; CHECK-DAG:   }
-    ; CHECK:     }
+    ; CHECK-DAG:       "onednn_conv_config":{$0$1
+    ; CHECK-DAG:    }
+    ; CHECK:      }
     )";
 
-  const char* fused_convolution_binary_add_ = R"(
-    ; CHECK:     custom_call_target="__onednn$convolution",
-    ; CHECK:       backend_config={
-    ; CHECK-DAG:     "outer_dimension_partitions":[],
-    ; CHECK-DAG:       "onednn_conv_config":{
-    ; CHECK-DAG:       "fusions":{
-    ; CHECK-DAG:         "ops":["BINARY_ADD"]
-    ; CHECK-DAG:     }
-    ; CHECK-DAG:   }
-    ; CHECK:     }
-    )";
-};
+  constexpr static char* kConvRewriteFusionsStr = R"(
+    ; CHECK-DAG:          "fusions":{
+    ; CHECK-DAG:            "ops":[$0]
+    ; CHECK-DAG:      },)";
```

Comment:
Nit: `kConvRewriteStr` and `kConvRewriteFusionStr` can be merged. We can also use named variables to improve readability.
```
  constexpr static char* kConvRewriteStr = R"(
    ; CHECK:     custom_call_target="__onednn$$convolution",
    ; CHECK:       backend_config={
    ; CHECK-DAG:     "outer_dimension_partitions":[],
    ; CHECK-DAG:       "onednn_conv_config":{
    ; CHECK-DAG:          "fusions":{
    ; CHECK-DAG:            "ops":[$fused_ops]
    ; CHECK-DAG:      },$opt_config
    ; CHECK-DAG:    }
    ; CHECK:      }
    )";
```

To substitute values, we can use [absl::StrReplaceAll](https://github.com/abseil/abseil-cpp/blob/8c495b5bbd54adeb7d82205c26830f76e50773b3/absl/strings/str_replace.h#L32-L37).

Reply:
In the absence of any fused ops the HLO string does not contain the empty `fusions` repeated field, hence I could not merge `kConvRewriteStr` and `kConvRewriteFusionStr`